embassy-stm32/src/timer/window_watchdog.rs

```diff
@@ -0,0 +1,149 @@
+use core::cell::RefCell;
+use core::future::poll_fn;
+use core::marker::PhantomData;
+use core::task::Poll;
+
+use embassy_cortex_m::peripheral::{PeripheralMutex, PeripheralState, StateStorage};
+use embassy_sync::waitqueue::WakerRegistration;
+use embassy_time::Duration;
+use stm32_metapac::timer::vals::CcmrInputCcs;
+
+use crate::gpio::sealed::AFType;
+use crate::pwm::*;
+use crate::Peripheral;
+
+/* requires VALID_THRESHOLD edges within range before reporting ok */
+const VALID_THRESHOLD: u8 = 2;
+
+pub struct State<'a, T: CaptureCompare16bitInstance>(StateStorage<StateInner<'a, T>>);
+impl<'a, T: CaptureCompare16bitInstance> State<'a, T> {
+    pub const fn new() -> Self {
+        Self(StateStorage::new())
+    }
+}
+
+struct StateInner<'a, T: CaptureCompare16bitInstance> {
+    phantom: PhantomData<&'a mut T>,
+    waker: WakerRegistration,
+    min: u16,
+    count: u8,
+}
+unsafe impl<'a, T: CaptureCompare16bitInstance> Send for StateInner<'a, T> {}
+
+pub struct WindowWatchdog<'a, T: CaptureCompare16bitInstance> {
+    inner: RefCell<PeripheralMutex<'a, StateInner<'a, T>>>,
```

Comment:
Could you not use `PeripheralMutex`? We're trying to move away from using it, I'd really like to not introduce new uses. You can pass a pointer to the irq with [set_handler_context](https://docs.embassy.dev/embassy-cortex-m/git/thumbv8m.main-none-eabihf/interrupt/trait.InterruptExt.html#tymethod.set_handler_context), then make that struct usable with `&` instead of `&mut`. (using atomics, mutexes, AtomicWaker etc)